src/jit/morph.cpp

```diff
@@ -7539,86 +7779,41 @@ void Compiler::fgMorphTailCall(GenTreeCall* call, void* pfnCopyArgs)
         // materialize as embedded stmts in right execution order.
         assert(thisPtr != nullptr);
         call->gtCallArgs = gtNewListNode(thisPtr, call->gtCallArgs);
+        // We changed args so recompute info.
+        call->fgArgInfo = nullptr;
     }
 
-#if defined(_TARGET_AMD64_)
-
-    // Add the extra VSD parameter to arg list in case of VSD calls.
-    // Tail call arg copying thunk will move this extra VSD parameter
-    // to R11 before tail calling VSD stub. See CreateTailCallCopyArgsThunk()
-    // in Stublinkerx86.cpp for more details.
-    if (call->IsVirtualStub())
+    // We may need to pass the target address, for instance for calli.
+    if (help.flags & CORINFO_TAILCALL_STORE_TARGET)
     {
-        GenTree* stubAddrArg = fgGetStubAddrArg(call);
-
-        // We don't need this arg to be in the normal stub register, so
-        // clear out the register assignment.
-        assert(stubAddrArg->gtRegNum == virtualStubParamInfo->GetReg());
-        stubAddrArg->gtRegNum = REG_NA;
-
-        // And push the stub address onto the list of arguments
-        call->gtCallArgs = gtNewListNode(stubAddrArg, call->gtCallArgs);
+        noway_assert(call->gtCallType == CT_INDIRECT && call->gtCallAddr != nullptr);
+        GenTree* target = call->gtCallAddr;
+        // TODO: This should be evaluated last. How?
+        call->gtCallArgs = gtNewListNode(target, call->gtCallArgs);
+        call->fgArgInfo = nullptr;
     }
```

Comment:
@dotnet/jit-contrib Here I want to change an indirect call to a call that puts the target address as the first argument. However the target expression needs to be evaluated last, not first. Any ideas on how I can accomplish this? The previous code kept the call as an indirect call and then did the transformation in lowering, but that seems pretty hacky.

Reply:
But is this target address a real argument? I see that you did not null out `gtCallAddr` nor changed the call type from `GT_INDIRECT`. That's unlikely to work properly as `gtCallAddr` and a few other call "operands" have fixed evaluation order and are generally treated differently from args.

Reply:
Ah, I see that the call type was changed below to `CT_USER_FUNC`. In which case yeah, this is a fun problem.

Reply:
I'm not actually sure the JIT does it right as it is, or maybe it guarantees the execution order somewhere else. This is the code responsible in lowering:
https://github.com/dotnet/coreclr/blob/b0df78cff9e2a1a0bbb669420a120444657913b2/src/jit/lower.cpp#L2181-L2222
It overwrites the second arg with the target address and afterwards changes it to a helper call. It looks to me like this will start evaluating `callTarget` as the second arg instead of after all args?

Reply:
Are these indirect calls produced by importing `calli`? It looks like the importer spills the address in `impImportIndirectCall`, that would provide more code motion freedom. Though I'm not sure if that handles the case of address being a variable that's somehow assigned during normal call args evaluation.

Reply:
The current mechanism turns any tail-prefixed call that cannot be dispatched directly into a call to `JIT_TailCall` where the original target is passed as the second arg. So it is used generally, not just for `calli`. However morph will explicitly generate a null-check in `fgMorphTailCallViaHelper` if required by the call, so that might alleviate it for all other cases than `calli` (not sure if, for virtual calls, this is the only side-effect the target expression could have).

Reply:
> However, here the evaluation order hasn't yet been set. I think that the only real alternative is to identify this case in gtSetEvalOrder and ensure that the first arg (the call target) is evaluated last in that case.

I'm not sure if `gtSetEvalOrder` would work since the order of the call args is actually determined by `SortArgs`. It may be problematic to have `gtSetEvalOrder` go against `SortArgs`.

To me it's not clear if reordering is an actual problem here. I've been trying to construct various `calli` trees but it seems that due to importer spilling it's unlikely to end up having side effects in the call address.

Reply:
So perhaps the alternatives are to address this in `SortArgs`, or to ensure that importer will spill in any case that needs this, and then perhaps add an assertion check for conflicting side effects.

Reply:
Does that seem correct @CarolEidt @mikedn? Also @erozenfeld maybe you have an idea too?

Reply:
That seems correct to me - that is, if it's not a simple lclVar it will spill it (and will spill anything with side effects as well). So adding that assert seems reasonable.

Reply:
I'm not sure the spilling done in the importer is sufficient. What if the call address itself has no side effects but one of the args has a side effect affecting the call address value? I don't think in this case impAppendStmt will spill that arg before the call address.

Reply:
@erozenfeld - it appears that you're right because it's not calling `impSpillStackEnsure` or otherwise ensuring that there are no side-effects in the stack, but I'm confused because it actually generates a separate statement (rather than inserting the assignment of the call address temp as a comma), so it would evaluate before the call evaluates any of its args.

Reply:
> What if the call address itself has no side effects but one of the args has a side effect affecting the call address value? I don't think in this case impAppendStmt will spill that arg before the call address.

Doesn't the expression get turned into an assignment to a temp? So it always has a side effect?

Reply:
The side effect that is due to the assignment to a local is ignored:
https://github.com/dotnet/coreclr/blob/f5437ac095272f1ab5e8ca79960add974fa96f2d/src/jit/importer.cpp#L560-L570

Reply:
I see, ok. @erozenfeld also mentioned that this could be a problem for any indirect call, not just calli. For example in an expression like
`tail. instance.Foo(instance = x);`
we could have the same problem. So we will have to find something out.
One not-so-elegant solution would be to change the stubs so that the target address is passed last. It should solve the problem but it is a hoax.

Reply:
A probably-even-less-elegant solution would be to spill all the arguments to temps. This is currently done in `fgArgInfo::EvalArgsToTemps`, called from `fgMorphArgs` when `fgArgInfo->NeedsTemps()` is true. I think it only spills the ones that are marked with `needTmp`.

Reply:
I think it might be reasonable. Perf wise it should not be a big problem as the target address is needed only rarely:
1. For `tail. calli`
2. For tailcalls to generic methods

For the second case we also do not have to do this for direct calls since the target expression will just be a function pointer. So really the only case this is needed is calli and generic virtual methods, and at least GVM's are already quite slow.

Reply:
> but I'm confused because it actually generates a separate statement (rather than inserting the assignment of the call address temp as a comma), so it would evaluate before the call evaluates any of its args.

I am also still confused about this, but I have a hard time coming up with IL where the arguments have non trivial side effects. In theory if we meet a `calli` in the importer while having the following stack we should see the problem:
```
GT_LCL_VAR x  // target
GT_ASG        // arg1
  GT_LCL_VAR x
  GT_CNS_INT 0
```
since, as @CarolEidt mentioned, it appears the importer will spill `target` to a statement that is evaluated before the assignment. Maybe someone else has an idea?

Reply:
I have been able to convince myself that making the call target the last arg in the stub might not actually be inelegant. It solves this problem, but it also makes sense since the arg is usually not there (so kind of optional) and it matches `calli`/`callvirt` behavior for null checks.